src/Illuminate/Database/Eloquent/Concerns/HasRelationships.php

```diff
@@ -394,7 +394,11 @@ protected function guessBelongsToRelation()
      * @return (
      *     $relationship is string
      *     ? \Illuminate\Database\Eloquent\PendingHasThroughRelationship<\Illuminate\Database\Eloquent\Model, $this>
-     *     : \Illuminate\Database\Eloquent\PendingHasThroughRelationship<TIntermediateModel, $this>
+     *     : (
+     *          $relationship is \Illuminate\Database\Eloquent\Relations\HasMany<TIntermediateModel, $this>
+     *          ? \Illuminate\Database\Eloquent\PendingHasThroughRelationship<TIntermediateModel, $this, \Illuminate\Database\Eloquent\Relations\HasMany<TIntermediateModel, $this>>
+     *          : \Illuminate\Database\Eloquent\PendingHasThroughRelationship<TIntermediateModel, $this, \Illuminate\Database\Eloquent\Relations\HasOne<TIntermediateModel, $this>>
+     *     )
```

Comment:
You can probably clean up the conditions if you add a Relation template type here:

```php
    /**
     * Create a pending has-many-through or has-one-through relationship.
     *
     * @template TIntermediateModel of \Illuminate\Database\Eloquent\Model
     * @template TLocalRelation of \Illuminate\Database\Eloquent\Relations\HasOneOrMany<TIntermediateModel, covariant $this>
     *
     * @param  string|TLocalRelation  $relationship
     * @return (
     *     $relationship is string
     *     ? \Illuminate\Database\Eloquent\PendingHasThroughRelationship<\Illuminate\Database\Eloquent\Model, $this>
     *     : \Illuminate\Database\Eloquent\PendingHasThroughRelationship<TIntermediateModel, $this, TLocalRelation>
     * )
     */
```

Reply:
Wow, thanks for your quick feedback!

I must admit this was my first idea as well, but somehow I couldn't get it to play right. Using this I was faced with:
```
 ------ ------------------------------------------------------------------------------------------------------------------------------
  Line   Relations.php
 ------ ------------------------------------------------------------------------------------------------------------------------------
  195    Expected type
         Illuminate\Database\Eloquent\PendingHasThroughRelationship<Illuminate\Types\Relations\Mechanic,
         $this(Illuminate\Types\Relations\User),
         Illuminate\Database\Eloquent\Relations\HasOne<Illuminate\Types\Relations\Mechanic,
         $this(Illuminate\Types\Relations\User)>>, actual:
         Illuminate\Database\Eloquent\PendingHasThroughRelationship<Illuminate\Types\Relations\Mechanic,
         $this(Illuminate\Types\Relations\User),
         Illuminate\Database\Eloquent\Relations\HasOneOrMany<Illuminate\Types\Relations\Mechanic, covariant
         $this(Illuminate\Types\Relations\User)>>
  199    Expected type Illuminate\Database\Eloquent\Relations\HasOneThrough<Illuminate\Types\Relations\Car,
         Illuminate\Types\Relations\Mechanic, $this(Illuminate\Types\Relations\User)>, actual:
         Illuminate\Database\Eloquent\Relations\HasManyThrough<Illuminate\Types\Relations\Car,
         Illuminate\Types\Relations\Mechanic,
         $this(Illuminate\Types\Relations\User)>|Illuminate\Database\Eloquent\Relations\HasOneThrough<Illuminate\Types\Relations\Car,
         Illuminate\Types\Relations\Mechanic, $this(Illuminate\Types\Relations\User)>
  215    Expected type
         Illuminate\Database\Eloquent\PendingHasThroughRelationship<Illuminate\Types\Relations\Mechanic,
         $this(Illuminate\Types\Relations\User),
         Illuminate\Database\Eloquent\Relations\HasMany<Illuminate\Types\Relations\Mechanic,
         $this(Illuminate\Types\Relations\User)>>, actual:
         Illuminate\Database\Eloquent\PendingHasThroughRelationship<Illuminate\Types\Relations\Mechanic,
         $this(Illuminate\Types\Relations\User),
         Illuminate\Database\Eloquent\Relations\HasOneOrMany<Illuminate\Types\Relations\Mechanic, covariant
         $this(Illuminate\Types\Relations\User)>>
  224    Expected type Illuminate\Database\Eloquent\Relations\HasManyThrough<Illuminate\Types\Relations\Car,
         Illuminate\Types\Relations\Mechanic, $this(Illuminate\Types\Relations\User)>, actual:
         Illuminate\Database\Eloquent\Relations\HasManyThrough<Illuminate\Types\Relations\Car,
         Illuminate\Types\Relations\Mechanic,
         $this(Illuminate\Types\Relations\User)>|Illuminate\Database\Eloquent\Relations\HasOneThrough<Illuminate\Types\Relations\Car,
         Illuminate\Types\Relations\Mechanic, $this(Illuminate\Types\Relations\User)>
 ------ ------------------------------------------------------------------------------------------------------------------------------
```

This seems to suggest that somehow `TLocalRelation` is not properly inherited as either `HasOne` or `HasMany`, but keeps using `HasOneOrMany` which is kind of unfortunate.